src/foam/nanos/dig/HTTPDigestSink.java

```diff
@@ -79,49 +78,46 @@ public void put(Object obj, Detachable sub) {
     FObject fobj = (FObject) obj;
     Object id = fobj.getProperty("id");
     String className = fobj.getClass().getSimpleName();
+    int responseCode = -1;
+    int retryCount = 3;
 
-    try {
-      int responseCode = sendRequest(fobj);
-
-      if ( responseCode == HttpServletResponse.SC_OK ) return;
-
-      if ( responseCode == HttpServletResponse.SC_BAD_REQUEST ) { // client error
-        String name = "HTTP DIGEST 400 RESPONSE";
-        String note = "[" + className + ", " + id + ", " + new Date() + "]";
-        createAlarm(name, note, LogLevel.WARN);
-
-      } else if ( responseCode == HttpServletResponse.SC_INTERNAL_SERVER_ERROR )  { // server error
-        try {
-          Thread.sleep(5000);
-        } catch (InterruptedException e) {}
-
-        // make a new request
+    while ( retryCount > 0 ) {
+      try {
         responseCode = sendRequest(fobj);
 
-        if ( responseCode == HttpServletResponse.SC_OK ) return;
+        // resend on 4xx and 5xx
+        if ( responseCode >= 400 && responseCode <= 599 )  {
 
-        String name = "HTTP DIGEST " + responseCode + " RESPONSE";
+          try {
+            Thread.sleep(5000);
+          } catch (InterruptedException e) {}
+
+          retryCount--;
+        } else {
+          return;
+        } 
+      } catch (SocketException e) {
+        // create an alarm on connection timeout
+        String name = "HTTP DIGEST CONNECTION TIMEOUT";
         String note = "[" + className + ", " + id + ", " + new Date() + "]";
-        createAlarm(name, note, LogLevel.WARN);
+        createAlarm(name, note, LogLevel.ERROR);
 
-      } else if ( responseCode == HttpServletResponse.SC_GATEWAY_TIMEOUT ) {
-        String name = "HTTP DIGEST 504 RESPONSE";
+        throw new RuntimeException(e);
+      } catch (Throwable t) {
+        String name = "HTTP DIGEST UNEXPECTED FAILURE";
         String note = "[" + className + ", " + id + ", " + new Date() + "]";
-        createAlarm(name, note, LogLevel.WARN);
-      }
+        createAlarm(name, note, LogLevel.ERROR);
 
-      throw new RuntimeException(this.getClass().getSimpleName() + "[" + className + ", " + id + ", " + responseCode + "]");
+        throw new RuntimeException(t);
```

Comment:
Whats the benefit of throwing RuntimeException here?

Reply:
sendRequest might fail for the reason other than socket error .. this catches that here

Reply:
and I believe all the exceptions thrown in the put are handled in the parent method